crates/bevy_math/src/cubic_splines.rs

```diff
@@ -846,11 +830,287 @@ impl<P: Point> IntoIterator for CubicCurve<P> {
+    /// length of the curve from start (t = 0) to end (t = n), where `n = self.segment_count()`,
+    /// returning an iterator evaluating the curve with the supplied `sample_function` at each `t`.
+    ///
+    /// For `subdivisions = 2`, this will split the curve into two lines, or three points, and
+    /// return an iterator with 3 items, the three points, one at the start, middle, and end.
+    #[inline]
+    pub fn iter_samples<'a, 'b: 'a>(
+        &'b self,
+        subdivisions: usize,
+        mut sample_function: impl FnMut(&Self, f32) -> P + 'a,
+    ) -> impl Iterator<Item = P> + 'a {
+        self.iter_uniformly(subdivisions)
+            .map(move |t| sample_function(self, t))
+    }
+
+    /// An iterator that returns values of `t` uniformly spaced over `0..=subdivisions`.
+    #[inline]
+    fn iter_uniformly(&self, subdivisions: usize) -> impl Iterator<Item = f32> {
+        let segments = self.segments.len() as f32;
+        let step = segments / subdivisions as f32;
+        (0..=subdivisions).map(move |i| i as f32 * step)
+    }
+
+    /// The list of segments contained in this `RationalCurve`.
+    ///
+    /// This spline's global `t` value is equal to how many segments it has.
+    ///
+    /// All method accepting `t` on `RationalCurve` depends on the global `t`.
+    /// When sampling over the entire curve, you should either use one of the
+    /// `iter_*` methods or account for the segment count using `curve.segments().len()`.
+    #[inline]
+    pub fn segments(&self) -> &[RationalSegment<P>] {
+        &self.segments
+    }
+
+    /// Iterate over the curve split into `subdivisions`, sampling the position at each step.
+    pub fn iter_positions(&self, subdivisions: usize) -> impl Iterator<Item = P> + '_ {
+        self.iter_samples(subdivisions, Self::position)
+    }
+
+    /// Iterate over the curve split into `subdivisions`, sampling the velocity at each step.
+    pub fn iter_velocities(&self, subdivisions: usize) -> impl Iterator<Item = P> + '_ {
+        self.iter_samples(subdivisions, Self::velocity)
+    }
+
+    /// Iterate over the curve split into `subdivisions`, sampling the acceleration at each step.
+    pub fn iter_accelerations(&self, subdivisions: usize) -> impl Iterator<Item = P> + '_ {
+        self.iter_samples(subdivisions, Self::acceleration)
+    }
+
+    #[inline]
+    /// Adds a segment to the curve
+    pub fn push_segment(&mut self, segment: RationalSegment<P>) {
+        self.segments.push(segment);
+    }
+
+    /// Returns the [`RationalSegment`] and local `t` value given a spline's global `t` value.
+    #[inline]
+    fn segment(&self, t: f32) -> (&RationalSegment<P>, f32) {
+        if self.segments.len() == 1 {
+            (&self.segments[0], t)
+        } else {
+            let i = (t.floor() as usize).clamp(0, self.segments.len() - 1);
+            (&self.segments[i], t - i as f32)
+        }
+    }
```

Comment:
Segments should be determined by spans in the knot vector.

Each successive pair of knots k_i, k_i+1 represents the interval [k_i, k_i+1) on which the input t maps to piecewise segment i. In the event than k_i = k_i+1, the span is zero and we skip segment i.

We have rationality, but we still have to get non-uniformity right.

Reply:
Ah, I thought the matrix already encoded that information.